### DISCUSSION TITLE
Migrate Django existing users with password into Keycloak

### DISCUSSION THREAD

**Discussion by MrArabboy** [keycloak/keycloak#17684] [Q&A] [answered]
Hello! I can not migrate django user password corto keycloak. I send data  like 
`{
            "email": user.email,
            "username": user.email,
            "enabled": True,
            "credentials": [
                {
                     "algorithm": "pbkdf2_sha256",
                     "type": "password",
                     "hashedSaltedValue": hashvalue,
                     "hashIterations": 320000,
                     "salt": salt,
                  }
            ]
        }
    `
but I can not login in Keycloak with  django password

**Comment by WalkerWalker** [comment]
I use this https://github.com/opensingular/singular-keycloak-database-federation
essentially this extension reads the django user pool from postgres and do the verification.
I would love to see other ways. Our system has a big chunk of Django legacy code and we are peel it off little by little.

**Comment by MrArabboy** [comment, marked as answer]
I found solution: 
`algorithm, iterations, salt, hashed_value = user.password.split("$")`
 and send payload like below:
 `{
            "email": user.email,
            "username": user.email,
            "enabled": True,
            "credentials": [
                {
                    "algorithm": algorithm.replace("_", "-"),
                    "type": "password",
                    "hashedSaltedValue": hashed_value,
                    "hashIterations": iterations,
                    "salt": base64.b64encode(salt.encode()).decode("ascii").strip(),
                }
            ]
        }`

**Reply by FVitor7** [reply]
and how can we return the password hash to django?

**Reply by hadpro24** [reply]
Thank you !!!
